### PR TITLE
doc for symbol attributes and naming convention (#2070)

### PR DESCRIPTION
**Pull request by pluskid**
@piiswrong I changed `list_attr()` to be shallow by default and added a function `list_attr_recursive()`. Also changed the separator between symbol name and attribute name to be `/` (instead of `_`).

@vchuravy What do we need to modify on the Julia side?

### CODE REVIEW

**Comment by piiswrong** [comment]
can we use list_attr(recursive=True) for recursive and set it to false by default?

File: docs/packages/python/symbol.md [hunk 1]
```diff
@@ -38,6 +39,45 @@ The following code gives an example of computation graph that add two inputs tog
...
+                          num_filter=1, attr={'mood': 'so so'})
+```
+Both key and values of the attribute dictionary should be strings, in order to properly communicate with the C++ backend. The attributes can be retrived via `attr(key)` or `list_attr()`:
+```
+assert data.attr('mood') == 'angry'
+assert op.list_attr() == {'mood': 'so so'}
+```
+In the case of a composite symbol, you can also retrieve all the attributes associated with that symbol *and its descendents* via `list_attr_recursive()`. Note in the returned dictionary, all the attribute names are with a prefix `'symbol_name' + '/'` in order to avoid naming conflicts.
+```python
+assert op.list_attr_recursive() == {'data/mood': 'angry', 'conv/mood': 'so so',
```
Comment (piiswrong) [inline docs/packages/python/symbol.md]: explain why mood is copied to conv_weight/mood here. also note that it won't happen if data is created by hand (is this true?) or if data is output of another symbol

File: docs/packages/python/symbol.md [hunk 2]
```diff
@@ -38,6 +39,45 @@ The following code gives an example of computation graph that add two inputs tog
...
+with mx.AttrScope(group='4', data='great'):
+    fc1 = mx.symbol.Activation(data, act_type='relu')
+    with mx.AttrScope(init_bias='0.0'):
+        fc2 = mx.symbol.FullyConnected(fc1, num_hidden=10, name='fc2')
+assert fc1.attr('data') == 'great'
+assert fc2.attr('data') == 'great'
+assert fc2.attr('init_bias') == '0.0'
+```
+
+**Naming convention**: it is recommended to choose the attribute names to be valid variable names. Names with double underscope (e.g. `__shape__`) are reserved for internal use. The slash `'/'` is the character used to separate a symbol name and its attributes, so should *not* be used in names of attributes.
```
Comment (piiswrong) [inline docs/packages/python/symbol.md]: add checks here to enforce not / is in name?

**Comment by piiswrong** [comment]
On a second thought this will actually cause problems with optimizer because tensors are named layername_weightname if you use _ as separator you'll have layername_weightname_lr_mult but if you use / it will be layername/weightname_lr_mult which is different from attaching lr_mult to variable directly

**Comment by pluskid** [comment]
@piiswrong No, you will actually have `layername_weightname/lr_mult`, at least in the case `weight` is created automatically by the operator (see the unit test). Let me check the case when user create an weight variable explicitly.

**Comment by pluskid** [comment]
Updated. @piiswrong I left out checking that the attributes name does not contain `/` because ~~I am too lazy~~ maybe sometimes we need to do some hacky stuff?

File: docs/packages/python/symbol.md [hunk 3]
```diff
@@ -38,6 +39,59 @@ The following code gives an example of computation graph that add two inputs tog
...
+                          num_filter=1, attr={'mood': 'so so'})
+```
+Both key and values of the attribute dictionary should be strings, in order to properly communicate with the C++ backend. The attributes can be retrived via `attr(key)` or `list_attr()`:
+```
+assert data.attr('mood') == 'angry'
+assert op.list_attr() == {'mood': 'so so'}
+```
+In the case of a composite symbol, you can also retrieve all the attributes associated with that symbol *and its descendents* via `list_attr(recursive=True)`. Note in the returned dictionary, all the attribute names are with a prefix `'symbol_name' + '/'` in order to avoid naming conflicts.
+```python
+assert op.list_attr_recursive() == {'data/mood': 'angry', 'conv/mood': 'so so',
```
Comment (piiswrong) [inline docs/packages/python/symbol.md]: also update doc

**Comment by piiswrong** [comment]
Ok. Please also update doc to reflect recusive=False change.
Otherwise LGTM

**Comment by piiswrong** [comment]
There is still the problem that you cannot use mx.sym.FullyConnected(attr={'weight_lr_mult': 0.1})

**Comment by pluskid** [comment]
@piiswrong Hmm... I see, this is indeed a problem....although in that way the conv operator will end up having (recursively) attributes like `weight_lr_mult`, `weight_weight_lr_mult`, `bias_weight_lr_mult`, etc.

**Comment by piiswrong** [comment]
Yes, it looks ugly but doesn't affect function.
Should we special treat this or leave it as _

**Comment by pluskid** [comment]
i guess best solution is to leave with _ for now until we find a better solution.

**Comment by pluskid** [comment]
Did not come up with a better solution. I changed `/` back to `_`. Let us get this merged for now.

**Comment by Godricly** [comment]
Hi, I was confused with online doc about that '_' in attribute naming. Do you mean the we should not use any '_' when naming an attribute? That 'init_bias' in the example is against my assumption. So maybe you mean not the first letter?

**Comment by pluskid** [comment]
@Godricly Thanks for pointing out. This was a mistake. I am fixing it in #2159.

**Pull request merged**